### DISCUSSION TITLE
Output tensor of Yolo11

### DISCUSSION THREAD

**Discussion by spiderwei** [ultralytics/ultralytics#17254] [General]
Could you please tell me where the object confidence score and class confidence score are in the output tensor of the yolo11 model?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @spiderwei, thank you for reaching out to Ultralytics! 🚀

We recommend visiting our [Docs](https://docs.ultralytics.com) for a comprehensive introduction, including examples of how to interpret YOLO model outputs and where you can find answers to commonly asked questions.

🌟 If your question relates to a bug or unexpected behavior, it would be very helpful if you could provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to assist us in diagnosing the issue.

🤔 For questions about custom training or model output interpretation, please supply as much detail as you can. This includes snippets of your code, model configurations, and any outputs you are encountering.

Join the Ultralytics community where you feel most comfortable. For lively discussions and support, check us out on [Discord](https://ultralytics.com/discord) 🎧. Prefer a forum format? Visit [Discourse](https://community.ultralytics.com). For deeper dives and knowledge sharing, our [Subreddit](https://reddit.com/r/ultralytics) is another great avenue.

## Upgrade

Ensure you have the latest version of `ultralytics` along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

Run YOLO in these supported environments, with all necessary dependencies preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. Check the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. Check the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing, verifying the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across different platforms every 24 hours and with each commit.

This response is automated, but an Ultralytics engineer will assist you soon. 😊

**Comment by glenn-jocher** [comment]
@spiderwei the YOLO11 output tensor includes object confidence scores and class probabilities. The object confidence is typically found in the first channel, while class probabilities follow. For detailed tensor structure, refer to the model documentation.

**Reply by spiderwei** [reply]
According to what you said, I made some debugging attempts on the model documentation, especially on the "tasks.py" file, but I didn't get the results I wanted. I know that the class probabilities in the yolov8 output tensor is after the image height value, that is, after the index number 3. So I want to ask，Is there any change in the index number of the object confidence  of yolo11 compared to yolov8?

**Reply by glenn-jocher** [reply]
@spiderwei in YOLO11, the object confidence score is typically located before the class probabilities in the output tensor, similar to YOLOv8. If you're encountering issues, please ensure you're using the correct model version and refer to the latest documentation for any updates.

**Reply by spiderwei** [reply]
Thank you for your reply. I have collected some data in the past few days, which are the output tensors of yolov8 and yolov5. I noticed that the output of yolov8 (batchsize, 84, 8400) is similar to the output of yolo11. But the "85" in the output of yolov5 (as shown below) contains the object confidence value, which is not available in the corresponding channels of the yolov8(yolo11) outputs. So I want to know where the target confidence missing from yolov8 and yolo11 is?  
<img width="747" alt="2" src="https://github.com/user-attachments/assets/91fcb204-0e03-4cd8-b6dc-2ebe4171b28e">
<img width="556" alt="1" src="https://github.com/user-attachments/assets/5dda91f1-0617-44bc-a9fc-177d8b8de953">
<img width="252" alt="3" src="https://github.com/user-attachments/assets/3a44bcbc-21b6-4378-b534-cf3fc52efa8a">

The first two pictures are the output tensor graphs of yolov8 and yolov5 that I collected, and the last one is the output tensor graph of yolo11 that I got through debugging (I set batchsize to 1).

**Reply by glenn-jocher** [reply]
@spiderwei in YOLO11, the object confidence is integrated into the class probabilities, similar to YOLOv8. The confidence score is derived from the combination of objectness and class probabilities, unlike YOLOv5 where it's explicitly separated.

**Reply by spiderwei** [reply]
What formula is used to combine object confidence and class probabilities into the class confidence score of yolo11? Can you tell me which file has the instructions? If you can help, I will be very grateful.

**Reply by glenn-jocher** [reply]
The class confidence score in YOLO11 is typically the product of the object confidence and class probabilities. For implementation details, please check the model's inference code in the `models` directory.

**Comment by vyrsh** [comment]
hello @glenn-jocher 
I am sorry, I tried to find the documentation on https://docs.ultralytics.com/models/yolo11/ and https://docs.ultralytics.com.

but I was not able to find the model documentation.

 I actually want to find the documentation for the output tensor. as I do not get what the 8400 stands for in the (1,84,8400).
 I know that the output tensor for a yolo model usually considers x,y of the top left corner. the height and width of the box. the confidence score and the objections. but this is 6, not 8400.
 
 thankyou for your assistance.

**Reply by glenn-jocher** [reply]
@vyrsh thank you for your question. The YOLO11 output tensor of shape `(1, 84, 8400)` can be understood as follows:

- **84**: Represents the number of channels, where the first 4 are for bounding box parameters (`x, y, width, height`), and the remaining 80 correspond to class probabilities for COCO's 80 classes.
- **8400**: Refers to the total number of grid points or anchor locations where predictions are made.

This structure enables YOLO to predict bounding boxes, object confidence, and class probabilities for multiple objects in an image. For further details, you can reference [YOLO11 documentation](https://docs.ultralytics.com/models/yolo11/). Let us know if you need additional clarification!

**Reply by varsand** [reply]
if my math is correct, does it not make it 85? 
x, y, w, h , object confidence, 80 classes = 85

**Reply by alex-quach-arc-boats** [reply]
There is no object confidence score. It took me several hours to realize this.
